2810 Move Judge Search to Elasticsearch

albertisfu opened the pull request:
@mlissner  this PR is still in progress. However, I wanted to describe how the approach of using Join fields for parent-child documents might change Judge search.

In #2856 I described how People model and its child objects. 

When it comes to querying Join documents in a parent-child relation in ES, we have two options.

**[Has parent queries:](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-child-query.html)**
Returns child documents whose joined parent document matches a provided query.

In this query, we can return child documents that match a query for all types (e.g., positions or education). Within each result, the parent document can be included as an inner hit.

- Education doc 1
  - Person 1
- Education doc 2
  - Person 1
- Education doc 3
  - Person 2
- Position 1
  - Person 1
 
 So for our use case where we require to show Judges (People) this approach it's not feasible.
 
**[Has child queries](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-parent-query.html):**
Returns parent documents whose joined child documents match a provided query.

Using this query type, we can retrieve parent documents (Judges) and within each result, include child documents that matched the query.

Thus, this approach is a better fit for Judge search. I've already implemented some methods for 'has child' queries in this PR, and the results look as follows:

e.g:
- Person 1
  - Education 1 that matched query.
- Person 2
  - Position 1 that matched query.
- Person 3
  - Education 2 that matched query.
  - Position 2 that matched query.
- Person 4 (no child objects matched)

However, there are some restrictions on this approach that can impact how search results look like. For instance, let's consider the following query: `Washington`,  results currently look like this in prod.

![Screenshot 2023-07-13 at 0 57 21](https://github.com/freelawproject/courtlistener/assets/486004/186a4ea0-a139-4f53-a13b-802ee40032bc)

It returns results where the term `Washington` matches either the Judge's name or any other searchable field (e.g., appointers).

Using `has child` queries, we'll also be able to return all the Judges that matched this query, although they might look a bit different.

Since a `has child` query returns parent documents and only those child documents that match the query, the results for this query might look as follows:

![Screenshot 2023-07-13 at 1 10 24](https://github.com/freelawproject/courtlistener/assets/486004/a5e490ae-4f92-478a-b23a-07dccc718361)

Notice that `Selection method` and `Appointers` are empty fields in `Washington, George Thomas Lane`, `George Washington`, while `Rutledge, John` does have `Appointers`.

The issue is that the first two results were matched by the Judge's name, and their positions fields don't match the term `Washington`, so they are not included within the result metadata.

Meanwhile, the third result, `Rutledge, John`, matches based on its `appointer` field in the child document position, so the position metadata is included in the result.

- A workaround for this behavior could be to perform an additional query for each result to return the positions and/or other child documents that are to be displayed on the search page.

- Alternatively, we could duplicate data from these fields in the parent document, so they can be displayed without performing an additional query.

- We could also alter the way results are displayed to take advantage of has child queries. This way, we can return results in a more structured manner.

For instance, we could return results that include the basic fields in the parent document, while child results that match the search can be included as a structured list within the same result.

e.g:
```
Rutledge, John (Supreme Court)
Born: September 18, 1739 in Charleston, South Carolina Deceased: June 21, 1800
...
Positions:
     Position type: "", Court: "", Selection Method: "Appointment (President)", Appointer:"Washington, George" 
     Position type: "", Court: "", Selection Method: "", Appointer:"" 
...
Education:
     School: "", Degree Level:"", Degree year: "" 
```
However if not child results are matched only the basic Judge info will be displayed.
```
Rutledge, John (Supreme Court)
Born: September 18, 1739 in Charleston, South Carolina Deceased: June 21, 1800
```

Let me know what do you think about this.

mlissner commented:
Great explanation, thank you. 

I think the best answer is to do the DB query for the missing data. That'll make for the richest search results and most consistent API results, and I can't imagine the performance would be much of a problem. It does feel like we'll need to have a new API endpoint for this though, since it'll be nested data, which we haven't served before.

albertisfu commented:
This PR is ready after @ERosendo updated it to use the `es_signal_processor,` made some improvements, and split the people tests into independent files.

So now, this PR serves as the base for our work on RECAP and Opinions search.

Here are some notes about it:

The final structure for Judge search is the following:

- PersonDocument (Parent document)
  - PositionDocument (Child document)

When I started working on this I tought `Education` was a good candidate to also become a child object. But then I realized we are only indexing one field from Education, `school.name`, so it's not worth creating a child object for it.

Since Join queries for parent-child documents are costly, we should use them only when required. In this case, `Position` is required since we need to index multiple fields from `Positions` and make them searchable.

In Solr, those fields are flattened and can't be searched independently from other positions.

e.g fields are flat in the Search API.
```
 "date_start": [
                "1972-10-17T07:53:00Z",
                "1943-01-01T07:53:00Z",
                "1952-01-01T07:53:00Z",
                "1959-01-01T07:53:00Z",
                "1965-01-01T07:53:00Z"
 ],
  "date_termination": [
      "2014-07-06T07:53:00Z",
      "1946-01-01T07:53:00Z",
      "1972-01-01T07:53:00Z",
      "1972-01-01T07:53:00Z",
      "1972-01-01T07:53:00Z"
  ]
```

Now that Position is a child object of `Person`, we'll be able to display positions that match a query in a nested structure within Judges.

e.g:
```

{
   "name":"Curtis Clinton",
   "name_reverse":"Clinton, Curtis",
   "....",
   "positions":[
      {
         "court_exact":"ca1",
         "position_type":"jud",
         "..."
      }{
         "court_exact":"ca2",
         "position_type":"jud",
         "..."
      }
   ]
}

```
However, this will require the creation of a new Search API endpoint (or a new version?) to display this new structure, since this is a breaking change. I created this [issue](https://github.com/freelawproject/courtlistener/issues/3033) to discuss the new search API endpoint.

We're using a [has_child](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-child-query.html) query to search within child objects. This query returns a list of parent documents, and only child documents that match the query are included within the results. The current Solr search for Judges returns all the positions for every Judge.

To maintain the current Solr Search API structure for Judges, and also the frontend search when moving to Elasticsearch, it was necessary to merge unavailable fields in search results from the database. This ensures that those fields are shown in the API and frontend. To do that one query for positions is performed in `merge_unavailable_fields_on_parent_document` for every page of results.

A significant advantage of using the `parent-child` approach and `has_child` queries over the current approach of flattening position fields is that users will be able to search over position fields as independent documents.

For example, let's consider the following documents and queries:

```
Judge 1
 Position 1:
   appointer: Clinton
   selection_method: e_part
  Position 2:
   appointer: Obama
   selection_method: a_legis

Judge 2:
 Position 1:
   appointer: Clinton
   selection_method: a_legis
```
   
Query:
```
{
            "type": SEARCH_TYPES.PEOPLE,
            "appointer": "clinton",
            "selection_method": "a_legis",
            "order_by": "name_reverse asc",
}
```

In the current Solr search this query will return `Judge 1` and `Judge 2`  as results since fields for positions are indexed as:

```
Judge 1
  appointer: ["Obama","Clinton"]
  selection_method: ["e_part", "a_legis"]
Judge 2
  appointer: ["Clinton"]
  selection_method: ["a_legis"]
```
So both documents are matched due to `appointer` and `selection_methods` are flat.

Using the new parent-child approach, this query will return only `Judge 2`, since the position fields are queried separately in each document.

- Highlighting has also been fixed for Judge search. Currently, judges names are not highlighted in Solr.  The problem is that in frontend the judge `name` is displayed using the `name_reverse` field, but search and filtering are performed using the `name` field. To fix this issue in ES, highlights in the `name` field are assigned to the `name_reverse` field, word by word, in order to highlight the correct terms. This transformation is performed in the `swap_fields_to_highlight` method. A similar transformation is performed to highlight the `dob_state` field, which is searched as an ID `dob_state_id`, but the results show the state name.

mlissner commented:
This sounds great, I'll review shortly. Two things before I begin. First:

> To fix this issue in ES, highlights in the name field are assigned to the name_reverse field, word by word, in order to highlight the correct terms. 

We can just show the names the same way in the front end. That's fine too.

Second: I just want to make sure v3 of the API is unchanged in terms of output, right? It'll just return better accuracy now, thanks to the parent-child approach?

albertisfu commented:
> We can just show the names the same way in the front end. That's fine too.

Yeah, this is already fixed in the frontend, so we show `name` highlights over the `name_reverse` field which is the one shown in Judge search results in frontend. In the API we only show highlights in the snippet.

> Second: I just want to make sure v3 of the API is unchanged in terms of output, right? It'll just return better accuracy now, thanks to the parent-child approach?

Yes, this is right, in terms of output the V3 Search API is the same, that's why we're merging fields from the DB that are not available in ES results due to `has_child` queries only return positions that matched the query and V3 API show all the positions data in flat fields for each judge. 

Yeah the improvement due to the `parent-child` approach is that now positions objects could be queried and filtered independently. The future improvement regarding API is to develop the [v4](https://github.com/freelawproject/courtlistener/issues/3033#issuecomment-1692432342) so we can show nested positions for each Judge.

mlissner commented on cl/search/api_serializers.py:
```diff
@@ -310,17 +311,71 @@ def create(self, validated_data):
+                "races": serializers.ListField(read_only=True),
+                "school": serializers.ListField(read_only=True),
+                "political_affiliation": serializers.ListField(read_only=True),
+                "political_affiliation_id": serializers.ListField(
+                    read_only=True
+                ),
+                "position_type": serializers.ListField(read_only=True),
+                "supervisor": serializers.ListField(read_only=True),
+                "predecessor": serializers.ListField(read_only=True),
+                "date_nominated": serializers.ListField(read_only=True),
+                "date_elected": serializers.ListField(read_only=True),
+                "date_recess_appointment": serializers.ListField(
+                    read_only=True
+                ),
+                "date_referred_to_judicial_committee": serializers.ListField(
+                    read_only=True
+                ),
+                "date_judicial_committee_action": serializers.ListField(
+                    read_only=True
+                ),
+                "date_hearing": serializers.ListField(read_only=True),
+                "date_confirmation": serializers.ListField(read_only=True),
+                "date_start": serializers.ListField(read_only=True),
+                "date_granularity_start": serializers.ListField(
+                    read_only=True
+                ),
+                "date_retirement": serializers.ListField(read_only=True),
+                "date_termination": serializers.ListField(read_only=True),
+                "date_granularity_termination": serializers.ListField(
+                    read_only=True
+                ),
+                "judicial_committee_action": serializers.ListField(
+                    read_only=True
+                ),
+                "nomination_process": serializers.ListField(read_only=True),
+                "selection_method": serializers.ListField(read_only=True),
+                "selection_method_id": serializers.ListField(read_only=True),
+                "termination_reason": serializers.ListField(read_only=True),
+            },
         }
```
If there's a way to avoid this too, that'd be great, but I don't know how easy it is.

ERosendo replied:
At first, I thought we were using this mapping because we didn't have a helper to identify the document fields and select a serializer class. However, this is not the case, we're using this mapping because those fields are not defined in the `ES Document`(We add all of those fields to the `PersonDocument` query results by using the `merge_unavailable_fields_on_parent_document` method).

I refactored this class to remove the mapping and use the DRF approach to [specify fields explicitly](https://www.django-rest-framework.org/api-guide/serializers/#specifying-fields-explicitly), so we can define serializer classes for documents like these:


```python
class PersonESResultSerializer(DocumentSerializer):
    """The serializer for Person results."""

    nomination_process = serializers.ListField(read_only=True)

    class Meta:
        document = PersonDocument
        exclude = ("text", "person_child")
```

In the example above, the class will serializer all the fields defined in the `PersonDocument` except for "text" and "person_child" and will also include "nomination_process" when deserializing objects.

mlissner commented on cl/search/api_utils.py:
```diff
@@ -43,8 +60,12 @@ def get_object_list(request, cd, paginator):
 
-    if cd["type"] == SEARCH_TYPES.ORAL_ARGUMENT and not waffle.flag_is_active(
-        request, "oa-es-deactivate"
+    if (
+        cd["type"] == SEARCH_TYPES.ORAL_ARGUMENT
+        and not waffle.flag_is_active(request, "oa-es-deactivate")
+    ) or (
+        cd["type"] == SEARCH_TYPES.PEOPLE
+        and not waffle.flag_is_active(request, "p-es-deactivate")
```
Isn't this:

```suggestion
if is_oral_argument_active or is_people_active:
```

?

ERosendo replied:
You're right!.

mlissner commented on cl/search/documents.py:
```diff
@@ -442,16 +441,26 @@ def prepare_political_affiliation(self, instance):
-        ]
+        ] or None
+
+    def prepare_text(self, instance):
+        text_template = loader.get_template("indexes/person_text.txt")
+        return text_template.render({"item": instance}).translate(null_map)
```
I have my same question here about whether we can get rid of this and still have a useful search engine. Seems like maybe?

ERosendo replied:
Yeah, we can remove this after adding a few fields to the `PersonDocument` class.

mlissner commented on cl/search/signals.py:
```diff
@@ -1,9 +1,21 @@
+from django.db.models.signals import m2m_changed, post_delete, post_save
+from django.dispatch import receiver
+from elasticsearch_dsl import Document
+
+from cl.alerts.send_alerts import send_or_schedule_alerts
 from cl.audio.models import Audio
+from cl.lib.command_utils import logger
+from cl.lib.elasticsearch_utils import es_index_exists
 from cl.lib.es_signal_processor import ESSignalProcessor
-from cl.people_db.models import Education, Person, Position
+from cl.people_db.models import (
+    ABARating,
+    Education,
+    Person,
+    PoliticalAffiliation,
+    Position,
+)
```
Lots of unused imports here.

ERosendo replied:
sorry, i'll remove them.

mlissner commented on cl/search/views.py:
```diff
@@ -488,7 +493,13 @@ def show_results(request: HttpRequest) -> HttpResponse:
                     ] = render_dict["search_summary_str"]
                     render_dict.update({"alert_form": alert_form})
-
+                elif request.GET.get("type") == SEARCH_TYPES.PEOPLE:
```
Can we factor out `request.GET.get("type")` at the top or something, please?

ERosendo replied:
sure!

mlissner reviewed (changes requested):
Everything looks pretty good here, though I'm not super happy with the many places that we're repeating field after field after field. It feels like it needs to be centralized or done dynamically or both. If we don't, it feels like adding a new field is going to be really painful down the road. 

Let me know what other questions I can help with. Overall, it's close!

ERosendo commented on cl/lib/es_signal_processor.py:
```diff
@@ -121,22 +118,31 @@ def save_document_in_es(
         if (
-            es_index_exists("people_db_index")
-            and parent_id
-            and PersonDocument.exists(id=parent_id)
+            not es_index_exists(es_document._index._name)
+            or not parent_id
+            or not instance.person.is_judge
         ):
```
Yes. I'll refactor this if statement. Thanks!

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -667,3 +1050,138 @@ def fetch_es_results(
+        queries_list.extend(
+            build_term_query(
+                "dob_state_id",
+                cd.get("dob_state", ""),
+            )
+        )
+        queries_list.extend(
+            build_term_query(
+                "political_affiliation_id",
+                cd.get("political_affiliation", "").split(),
+            )
+        )
+        queries_list.extend(
+            build_daterange_query(
+                "dob",
+                cd.get("born_before", ""),
+                cd.get("born_after", ""),
+            )
+        )
+        queries_list.extend(
+            build_text_filter(
+                "dob_city",
+                cd.get("dob_city", ""),
+            )
+        )
+        queries_list.extend(
+            build_text_filter(
+                "name",
+                cd.get("name", ""),
+            )
+        )
+        queries_list.extend(
+            build_text_filter(
+                "school",
+                cd.get("school", ""),
+            )
+        )
+
+        # Build position has child filter:
+        queries_list.extend(build_has_child_filters("position", cd))
```
There's no a reason to call the extend method six times. I'll refactor this method.

ERosendo replied:
We use the extend method six times because each `build_` helper returns a `list`. However, we can accomplish the same result with just one extend method.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -395,25 +447,107 @@ def add_es_highlighting(
     :return: The modified Elasticsearch search query object with highlights set
     """
 
+    types_to_highlight = [SEARCH_TYPES.ORAL_ARGUMENT, SEARCH_TYPES.PEOPLE]
+    if cd["type"] not in types_to_highlight:
+        return search_query
```
I'm sorry it's not clear. We do highlighting for everything. We use the method called `group_search_results` to do highlighting for `Parentheticals` and the method called `add_es_highlighting` for `Oral argument` and `people`. I will refactor the `build_es_main_query` method to make it clear.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -395,25 +447,107 @@ def add_es_highlighting(
+
+    elif cd["type"] == SEARCH_TYPES.PEOPLE:
+        highlighting_fields = SOLR_PEOPLE_ES_HL_FIELDS
+        hl_tag = SEARCH_HL_TAG
```
You're right! I'm refactoring this.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -395,25 +447,107 @@ def add_es_highlighting(
+    elif cd["type"] == SEARCH_TYPES.PEOPLE:
+        highlighting_fields = SOLR_PEOPLE_ES_HL_FIELDS
+        hl_tag = SEARCH_HL_TAG
+        fields_to_exclude = []
```
That's right.

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -229,13 +240,29 @@ def build_sort_results(cd: CleanData) -> Dict:
     """
 
-    if cd["type"] == SEARCH_TYPES.ORAL_ARGUMENT:
+    if (
+        cd["type"] == SEARCH_TYPES.ORAL_ARGUMENT
+        or cd["type"] == SEARCH_TYPES.PEOPLE
+    ):
```
> Couldn't you just have one order_by_map and put the parenthetical ones in the same big dict?

This is a great idea. We just have to change the value of the `"score desc"` key for `Parentheticals` 

> Would that allow you to remove the whole if/else?

Yes!

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -667,3 +1050,138 @@ def fetch_es_results(
+            score_mode="max",
+            query=reduce(operator.iand, queries_list),
+            inner_hits={"name": f"filter_inner_{child_type}"},
+            max_children=10,
```
Yes, the `max_children` attribute sets a limit to the number of **positions** we get for each document but We've just realized that If the parent document exceeds this limit, it is excluded from the search results (which is not desired) so we're going to remove this attribute and use the [size](https://www.elastic.co/guide/en/elasticsearch/reference/current/inner-hits.html#inner-hits-options) option instead. The docs says the following about the size option: 

> The maximum number of hits to return per inner_hits. By default the top three matching hits are returned.

mlissner replied:
I had to read this a bunch of times before getting it. `max_children` is the max number of children a parent can have and still show up in search results? Like, if there are 10 children and the max is five, then the whole thing is removed from search results???

ERosendo replied:
> max_children is the max number of children a parent can have and still show up in search results?

Yes

> if there are 10 children and the max is five, then the whole thing is removed from search results???

That's right. The document with 10 children wont show up in search results.

mlissner replied:
On what planet is that useful?

ERosendo commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -395,25 +447,107 @@ def add_es_highlighting(
+                replace_value_with_tag(
+                    highlights[field][0], original_value, field, tag
+                )
+            ]
```
understood! I'll tweak the template to show state IDs and the regular judge name in the UI.

albertisfu commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -466,6 +496,33 @@ def build_es_base_query(search_query: Search, cd: CleanData) -> Search:
+        case SEARCH_TYPES.PEOPLE:
+            child_query_fields = {
+                "position": add_fields_boosting(
+                    cd, ["appointer", "supervisor", "predecessor"]
```
I think here we need to add all the position fields that were present in `person_text.txt` so those fields can be searchable by the text query, e.g: `position_type_display`, `nomination_process_display`, `organization_name`, `court.pk`, etc.

ERosendo replied:
you're right! the `person_text.txt` file includes a few fields from the `position` document.

albertisfu commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -466,6 +496,33 @@ def build_es_base_query(search_query: Search, cd: CleanData) -> Search:
+                    "fjc_id",
+                    "aba_rating",
+                    "educations",
+                    "political_affiliation",
```
It seems `political_affiliation` is duplicated here. Maybe it should be `political_affiliation_id`? However I don't see that `political_affiliation_id`  was in `person_text.txt`

ERosendo replied:
You're right!. `political_affiliation` is duplicated and we don't need `political_affiliation_id`. I will fix the list.

albertisfu reviewed (commented):
@ERosendo these changes and refactors looks good to me. I just found a few issues, some comments below. 

Thanks!

ERosendo commented on cl/search/views.py:
```diff
@@ -476,25 +481,40 @@ def show_results(request: HttpRequest) -> HttpResponse:
+                search_type = request.GET.get("type")
+                match search_type:
+                    case SEARCH_TYPES.PARENTHETICAL:
+                        search_results = do_es_search(request.GET.copy())
+                        render_dict.update(search_results)
```
You're right! Thanks for bringing this to our attention

ERosendo commented on cl/search/views.py:
```diff
@@ -476,25 +481,40 @@ def show_results(request: HttpRequest) -> HttpResponse:
+                        render_dict.update(search_results)
+                    case SEARCH_TYPES.ORAL_ARGUMENT:
+                        # Check if waffle flag is active.
+                        if waffle.flag_is_active(request, "oa-es-active"):
+                            search_results = do_es_search(request.GET.copy())
+                        else:
+                            search_results = do_search(request.GET.copy())
+
+                        render_dict.update(search_results)
```
Sure!

ERosendo commented:
@albertisfu thanks for your review. After tweaking the code according to your comments, here is a summary of the changes: 

1. We refactored the `show_results` view to fix the indentation of the `match-case` statement and removes the `search_results` variable.
2. The commits removed the duplicate item (`political_affiliation`) from the list of fields to boosts.
3. We removed the item called `educations` and use `school` instead in the `add_fields_boosting` method.
4. We're adding the following field to the `Position` document class (we're using them to remove the `person_text.txt`):
    - court_full_name
    - court_id_text
    - court_citation_string
    - organization_name
    - job_title
    - position_type_text
    - judicial_committee_action_text
    - nomination_process_text
    - selection_method_text
    - termination_reason_text
  
5. We tweaked the mapping to add the **court's short name** to the query results. 
        
    We were adding the **full name** to the query results by using the `merge_unavailable_fields_on_parent_document` method but we noticed the current implementation of the API uses the **short name** instead so we changed it to preserve the API response exactly as it is.

mlissner commented:
Does this need to get rebased onto main before review? I think it might be kind of behind?

ERosendo commented:
@mlissner @albertisfu While I was working on the PR related to **Opinions**, I noticed the `DocumentSerializer` class was not working properly with fields defined as `fields.ListField`. I checked the [django_elasticsearch_dsl](https://github.com/django-es/django-elasticsearch-dsl) repo and noticed `ListField` is not a class like `KeywordField` or `TextField`. The `ListField` is a function that tweaks the implementation of the `get_value_from_instance` method  ( [Here's](https://github.com/django-es/django-elasticsearch-dsl/blob/f6fa886973c71415932500e4ca16519001429d31/django_elasticsearch_dsl/fields.py#L164-L177) the implementation of the helper method), this helper makes sense because there is no dedicated array data type in **Elasticsearch** but the `DocumentSerializer` is not able to identify which field contains arrays if we just use the `ListField` when we define the `Document` class..

I reviewed the [Field](https://github.com/elastic/elasticsearch-dsl-py/blob/bde8e10cbdefca654f402e077d8c885d182384f1/elasticsearch_dsl/field.py#L65) class from [elasticsearch-dsl-py](https://github.com/elastic/elasticsearch-dsl-py) and found that it has an attribute called `multi` to specify whether a field contains arrays of values or not, so we'll use this attribute in the `DocumentSerializer` to identify fields that should be serialized as lists so we need to add the `multi` attribute to those fields we want to serialize as lists. The following are some examples:

```python
# List of keywords:
field_1 = fields.ListField(fields.KeywordField(multi=True))  

# List of text:
field_2 = fields.ListField(
        fields.TextField(
            analyzer="text_en_splitting_cl",
            search_analyzer="search_analyzer",
            multi=True,
        )
    ) 

# List of integers:
field_3 = fields.ListField(fields.IntegerField(multi=True))
```

mlissner commented on cl/search/api_utils.py:
```diff
@@ -27,10 +30,24 @@ def get_object_list(request, cd, paginator):
+        request, "p-es-activate"
+    )
+
+    if is_oral_argument_active or is_people_active:
+        search_query = (
+            AudioDocument.search()
+            if is_oral_argument_active
+            else PersonDocument.search()
+        )
```
Call me crazy, but isn't it a lot easier to just do:

```suggestion
    if is_oral_argument_active:
        search_query = AudioDocument.search()
    elif is_people_active:
        search_query = PersonDocument.search()
```

Removes two lines, works the same, and is easier to read?

albertisfu replied:
Sure, changed it.

mlissner commented on cl/search/constants.py:
Feels like boosting could go in here too?

albertisfu replied:
Yeah, I moved BOOST to constants.

mlissner reviewed (dismissed):
Well, there's a whole lot of code here, and I've read it before, so I don't have a whole lot to say, but here are a few comments. 

If you guys are happy with this after these comments, I think it's time we take the leap.

albertisfu commented:
Some additional changes were required here:

**Indexing all positions for Judges:**

As we commented in #2856 a Judge and their positions are only indexed when a judiciary position is created for a person. This means that any non-judiciary positions created beforehand are not indexed. To address this, I've introduced `person_first_time_indexing`, which not only indexes the judiciary position and the judge for the first time but also retroactively indexes any non-judiciary positions that were previously skipped.

Aldo added tests for this method.

**Duplicate all the required Person fields into Positions:**

I've realized that we still need to duplicate some additional Person fields into Positions. Previously, we had duplicated all the essential Person fields to make both the filter and text queries functional. However, to enable all searchable fields in advanced queries, I've added the missing fields:
`id, alias_ids, races, political_affiliation_id, dob, dod, and dob_state_id`

So now these Person fields can be combined with Position fields in advance queries like: 
`appointer:Clinton AND political_affiliation_id:i`

The Person fields that are not duplicated in Positions are the ones that are not searchable:
`name_reverse, date_granularity_dob, date_granularity_dod, absolute_url`

I also can confirm that the current approach to filters and queries used for Judges is working as expected. Initially, I thought we might need to update the method for building filters and queries, similar to the approach I used in RECAP, especially after duplicating parent fields into child documents. However, this isn't currently necessary for Judges, as queries return the correct results when combining parent and child fields. So the current implementation can be merged as is. Once RECAP is merged, I can resolve any conflicts and update the method for building filters and queries.

The key difference is that, currently, in Judges, we do not display nested child objects in the results. Therefore, we only need to match parent objects if at least one child object matches the filters/queries. However, updating the approach to filters and queries when RECAP is merged will make Judges capable of displaying nested child items in the results, which will be necessary for the V4 Search API.

- I also fixed the `test_merge_unavailable_fields_api` test which was failing intermittently due to the random order of field values when merged from the database. I've modified the method to compare values regardless of their order.

- Added the new person fields into the signal mapping for updating person fields into position documents.

- Fixed a bug related to reverse objects. When a reverse object like a `PoliticalAffiliation` was removed from DB its related value was not updated in the related field in `Person` and `Position` documents. So I tweaked the `es_signal_processor` to include `delete_reverse_related_documents` in order to update the person and position fields when a reverse object is removed from the DB and added new tests related to this issue.

Writing this I was testing the process that we should follow for releasing this to prod and I found that the Django Elasticsearch DSL built-in commands:

```
search_index --populate --models people_db.Person --parallel
search_index --populate --models people_db.Position --parallel
```
Won't work for indexing all the existing Judges properly for two reasons:

- It will index everything including Persons who are not Judges.
- Positions are not properly linked to their parent document in the index.

So I'm afraid that we'll need a custom command for indexing documents that use the parent-child approach, I'll start checking what can we do about it.

albertisfu commented on cl/lib/search_utils.py:
```diff
@@ -55,15 +55,33 @@
+            "fjc_id": 1,
+            "aba_rating": 1,
+            "school": 1,
+            "political_affiliation": 1,
```
Sure, I removed all the fields with boost 1. Instead, they're added to the `query_string`  without boosting notation which according to [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#_boosting) they get the default value that is 1.

mlissner commented:
Thanks Alberto. This all sounds good to me. @ERosendo, will you please do a review and if you're both agreed, we'll merge. 

For the indexing question, we could do a hack. There aren't that many people that aren't judges. Could we just add everything then do some quick custom code to remove folks we don't want in the index?

ERosendo commented on cl/lib/es_signal_processor.py:
```diff
@@ -495,3 +582,37 @@ def handle_reverse_actions(self, sender, instance=None, **kwargs):
+            if isinstance(
+                instance, (ABARating, PoliticalAffiliation, Education)
+            ):
+                query_string_split = query_string.split("__")
+                query_string = query_string_split[-1]
+                instance_field = query_string_split[0]
```
I think we can refactor these three lines using tuple unpacking:

```python 
instance_field, *_, query_string = query_string.split("__")
```

ERosendo reviewed (approved):
The code looks good to me, just a small comment, but I think we can merge it.

mlissner commented:
Merging! Let's get the tweak into the upstream branch. Thank you guys!
